### DISCUSSION TITLE
Elegant way to check gitdiff files on PR for arm-estimator?

### DISCUSSION THREAD

**Discussion by riosengineer** [TheCloudTheory/arm-estimator#268] [General]
Hey, 

Wondering if anyone has gone through the pain of figuring out a clean and sleek way to only run the tooling on commits present in a PR within Azure DevOps? 

Would be awesome to run an estimation of costs for any files edited on PR, and even post the table in the ADO comments. 

For now, it looks like I will need to create a ton of custom logic to get this done using gitdiff, and then figure out the paired param file based on repo structure.

**Comment by kamil-mrzyglod** [comment]
Hey @riosengineer, do you need to run estimation for all the Bicep / ARM Templates files during PR, or only for the ones, which have changed? I don't know how the structure of your files looks like, but ACE is intended to run estimations against selected scope (resource group, subscription, etc.). In other words - it produces both total estimation and diff as long as ARM is able to provide what has changed. If this is your case, then it means, you don't need git diff - you will always run estimation for all the files and get the cost difference.

If the cost difference is not reported, then it's either ARM not returning the proper value, or there's a bug in ACE :)

**Reply by riosengineer** [reply]
My idea is to have it run only on the bicep files that have changes when someone creates a PR in ADO. This way we can see an estimated cost for only those files, then I would try and post them in the PR comments so we can see cost estimation for the changes an engineer is making. Just an idea at the moment, so was wondering if anyone had insights on how they have done that. 

If i have a structure like: 

```
└── src/ 
    ├── project/
    │   ├── dev/
    │   │   └── dev.bicepparam
    │   ├── test/
    │   │   └── test.bicepparam
    │   └── main.bicep
    └── other folders with same structure/
```

I basically have to run a gitdiff and find all changes bicep files in the PR, then run the arm-estimator for each of those. The challenge of also pairing it up with it's correct parameter file

**Reply by kamil-mrzyglod** [reply]
I think ADO doesn't have a native feature for that. All I can think about now is using REST API for doing the heavylifting and feeding the pipeline with proper output. https://learn.microsoft.com/en-us/rest/api/azure/devops/git/commits/get-changes?view=azure-devops-rest-5.0&tabs=HTTP#examples

**Reply by riosengineer** [reply]
Nope but you can use Git CLI to get all the files that have changed like `git diff HEAD HEAD~ --name-only -- bicep/* ":(exclude)*/.tests/*" | Where-Object {$_.EndsWith('.bicep')})`

I will see how I get on, if I can get it working & even posting a PR comment with the table, maybe it's something you would be interested in for the project.

**Reply by kamil-mrzyglod** [reply]
Sure, if you find a smart way to do that, happy to discuss that further :)

**Reply by riosengineer** [reply]
Getting somewhere :) On PR, gitdiff to find all changes bicep files and their corresponding param to run ACE with the output.
<img width="1439" alt="image" src="https://github.com/user-attachments/assets/f0248139-179c-4ccb-9043-dd6733f93b1d">
Posts the markdown output, formatting is a bit off, I am not sure if ADO comments expect certain markdown formatting. The markdown looks fine in PS terminal though. Will keep playing!

**Reply by kamil-mrzyglod** [reply]
Not sure if ADO accepts markdown as format for the comments. The markdown may be accepted only as input when writing a comment and transformed to HTML on the fly. Just a blind shot :)

**Reply by riosengineer** [reply]
I ended up using PowerShell `Get-Content -Path $markdownFile -Raw` and it looks pretty good I think!

<img width="1382" alt="image" src="https://github.com/user-attachments/assets/094c5de7-0f24-4594-895f-d57d609c69c8">

**Reply by kamil-mrzyglod** [reply]
Superb! Feel free to request changes to the output if you feel something's missing.
